add docs for query recorder (#141)
tensorchord/pgvecto.rs-docs: src/vectorchord/usage/measure-recall.md

$$, exact_search => true);
```

::: tip IMPORTANT
The `query` statement argument must start with `SELECT ctid`.
:::

## Query Recorder <badge type="tip" text="since v0.5.2" />

> In a production environment, data-driven continuous evaluation often requires insights from user-generated queries. While [log_statement](https://www.postgresql.org/docs/current/runtime-config-logging.html#GUC-LOG-STATEMENT) and [pgaudit](https://www.pgaudit.org/) can also capture real-time queries, filtering and parsing their logs often requires additional tools and infrastructure.

Author: OK, let's simplify the introduction to a sentence:

You may find it more useful to evaluate user-generated queries than randomly generated ones.

VectorChord includes a built-in query recorder to sample and store recent user queries. These records are grouped by index, making them easily accessible and seamlessly integrable with other functions like dump or recall evaluation.

```sql
-- Enable the query recorder
ALTER SYSTEM SET vchordrq.query_sampling_enable = on;

-- Set the maximum number of queries to record per index
ALTER SYSTEM SET vchordrq.query_sampling_max_records = 1000;

-- Set the sampling rate (e.g., 1.0 for all queries, 0.01 for 1%)
-- For high-traffic systems, a rate of 0.01 or lower is recommended.
ALTER SYSTEM SET vchordrq.query_sampling_rate = 1;
SELECT pg_reload_conf();

-- Now, execute a query against a VectorChord vchord index
-- SELECT * FROM items ORDER BY embedding <-> '[3,1,2]' LIMIT 10;
```

After queries have been executed and sampled, you can inspect the captured components of them, including the schema, table, column, operator, and vector.

```sql
-- List recorded queries for a specific index
SELECT * from vchordrq_sampled_queries('items_embedding_idx');
-- schema_name | table_name | column_name | operator | vector_text
---------------+------------+-------------+----------+-------------

> -- public      | items      | embedding   | <->      | [3,1,2]

Author: From where I stand, if the user want to focus on index name, it might be redundant for us to return what they input.

If they want to focus on table name, they may prefer to use `SELECT * from vchordrq_sampled_queries WHERE table_name='t';`.

Actually, the user doesn't really need index name for result, it's only for filter keys. By this way, we provide flexible options.

> -- public      | items      | embedding   | <->      | [3,1,2]

Reviewer: > Actually, the user doesn't really need index name for result, it's only for filter keys. By this way, we provide flexible options.

Why `SELECT * from vchordrq_sampled_queries WHERE table_name = 't'` is useful? This is the wrong way to code, and we should not encourage users to make mistakes.

-- List recorded queries for all accessible indexes
SELECT * from vchordrq_sampled_queries;
-- schema_name | table_name | column_name | operator | vector_text
---------------+------------+-------------+----------+-------------
-- public      | items      | embedding   | <->      | [3,1,2]
```

> To measure the recall of the recorded queries, you can reconstruct the original query statement.

Author: We will use use quote_literal and quote_ident to fix it.
ref: https://github.com/tensorchord/VectorChord/pull/334#discussion_r2331926998

```sql
SELECT AVG(recall_value) FROM (
    SELECT vchordrq_evaluate_query_recall(
            query => format(
                'SELECT ctid FROM %I.%I ORDER BY %I %s ''%s'' LIMIT 10',
                lq.schema_name,
                lq.table_name,
                lq.column_name,
                lq.operator,
                lq.vector_text
            )
    ) AS recall_value
    FROM vchordrq_sampled_queries('items_embedding_idx') AS lq
) AS eval_results;
```

::: tip

> - The feature will introduce a latency of approximately 1ms to SELECT statements.

Author: By discussion, 1ms is appraprate for most cases, or user can lower the query_sampling_max_records to get lower latency.

I will check the latency of 1600 dimension.

- Query records are not synchronized by replication or backup. Each Postgres replica maintains its own set of records based on the queries it processes.
- Every user can only access the recorded queries for indexes they have permission to access.
:::

## Reference

> This feature is not supported by `vchordg` and `MaxSim`.

Reviewer: I think this restriction doesn‘t make much sense; we’d better remove it as soon as possible.

### Search Parameters <badge type="info" text="vchordrq" /> {#search-parameters}

#### `vchordrq.query_sampling_enable` <badge type="tip" text="since v0.5.2" />

- Description: Enables or disables the sampling recorder of `vchordrq` index scan.
- Type: boolean
- Default: `off`

#### `vchordrq.query_sampling_max_records` <badge type="tip" text="since v0.5.2" />

- Description: Controls the maximum number of queries recorded per index. This setting is only active when `vchordrq.query_sampling_enable` is on.
- Type: integer
- Default: `0`
- Domain: `[0, 10000]`
- Example:
    - `vchordrq.query_sampling_max_records = 0` disables the recording of new queries and clears all existing records.
    - `vchordrq.query_sampling_max_records = 1000` will maintain up to the 1000 most recent queries for each index.

#### `vchordrq.query_sampling_rate` <badge type="tip" text="since v0.5.2" />

- Description: Controls the fraction of vchordrq index scan queries that are recorded. This setting is only active when `vchordrq.query_sampling_enable` is on.
- Type: float
- Default: `0`
- Domain: `[0, 1]`
- Example:

>     - `vchordrq.query_sampling_rate = 0` disables query recording.

Author: Let's change it to:
`vchordrq.query_sampling_rate = 0` prevents further queries from being recorded.

>     - `vchordrq.query_sampling_rate = 0` disables query recording.

Author: For GUC = 0, I will try to keep similar description.

    - `vchordrq.query_sampling_rate = 0.01` records approximately 1% of queries.
    - `vchordrq.query_sampling_rate = 1` records every query.

### Functions <badge type="info" text="vchordrq" />

#### `vchordrq_sampled_queries` <badge type="tip" text="since v0.5.2" /> {#vchordrq-sampled-queries-func}

- Description: Retrieves the set of recorded queries for a **single**, **specified** vchordrq index.
- Result: `table`
    - `schema_name: text`, the schema of the table in the query
    - `table_name: text`, the name of the table in the query
    - `column_name: text`, the column involved in the index scan
    - `operator: text`, the distance [operator](/vectorchord/usage/indexing#operator-classes) used in the query.
    - `vector_text: text`, the text representation of the target vector from the query
- Arguments:

>     - `regclass`, an object identifier of the `vchordrq` index

Reviewer: ```suggestion
    - `regclass`, the object identifier of a `vchordrq` index
```

- Example:
    - `SELECT vchordrq_sampled_queries('idx');`

#### `vchordrq_evaluate_query_recall` <badge type="tip" text="since v0.5.0" />

- Description: Evaluates the recall of a given `vchordrq` query.
- Result: `float`
- Arguments:
    - `query: text`, the query whose recall is evaluated
    - `exact_search: boolean`, use ground truth instead of estimated ground truth
    - `accu_probes: text`, `vchordrq.probes` for estimated ground truth

>     - `accu_epsilon: float`, `vchordrq.epsilon` for estimated ground truth

Reviewer: This is no such type in PostgreSQL.

- Example:
    - `SELECT vchordrq_evaluate_query_recall(query => $$ SELECT ctid FROM t ORDER BY val <-> '[0.5, 0.25, 1.0]' LIMIT 10 $$);`
    - `SELECT vchordrq_evaluate_query_recall(query => $$ SELECT ctid FROM t ORDER BY val <-> '[0.5, 0.25, 1.0]' LIMIT 10 $$, exact_search => true);`
    - `SELECT vchordrq_evaluate_query_recall(query => $$ SELECT ctid FROM t ORDER BY val <-> '[0.5, 0.25, 1.0]' LIMIT 10 $$, accu_probes => '9999', accu_epsilon => 4.0);`

### Views <badge type="info" text="vchordrq" />

#### `vchordrq_sampled_queries` <badge type="tip" text="since v0.5.2" />

- Description: Retrieves the set of recorded queries for **all** vchordrq indexes.

> - Result: `table`, same as the function [vchordrq_sampled_queries](#vchordrq-sampled-queries-func)

Reviewer: ```suggestion
```

$$, exact_search => true);
```

## Query sampling <badge type="tip" text="since v0.5.2" />

You may find it more useful to evaluate user-generated queries than randomly generated ones. VectorChord includes a built-in method to sample and store recent user queries. These records are grouped by index, making them easily accessible and seamlessly integrable with recall evaluation.

```sql
-- Enable the query sampling
ALTER SYSTEM SET vchordrq.query_sampling_enable = on;

-- Set the maximum number of queries to record per index
ALTER SYSTEM SET vchordrq.query_sampling_max_records = 1000;

-- Set the sampling rate (e.g., 1.0 for all queries, 0.01 for 1%)
-- For high-traffic systems, a rate of 0.01 or lower is recommended.
ALTER SYSTEM SET vchordrq.query_sampling_rate = 1;
SELECT pg_reload_conf();

-- Now, execute an example vector search
-- SELECT * FROM items ORDER BY embedding <-> '[3,1,2]' LIMIT 10;

> ```

Reviewer: 1. Referring to other documents, a descriptive paragraph is needed here.
2. Configuring parameters with a global setting is not intended. Please set GUC only when necessary.

After queries have been executed and sampled, you can inspect the captured components of them, including the schema, index, table, column, operator, and vector.

```sql

> -- List recorded queries for a specific index

Reviewer: Referring to other documents, it's best not to mix text with SQL statements.

SELECT * from vchordrq_sampled_queries('idx');

> -- schema_name | index_name | table_name | column_name | operator |    value
> ---------------+------------+------------+-------------+----------+--------------
> -- public      | idx        | items      | embedding   | <->      | [0.5,0.25,1]

Reviewer: ```suggestion
--  schema_name |     index_name      | table_name | column_name | operator |  value  
-- -------------+---------------------+------------+-------------+----------+---------
--  public      | items_embedding_idx | items      | embedding   | <->      | [3,1,2]
```

> -- List recorded queries for all accessible indexes

Reviewer: Referring to other documents, it's best not to mix text with SQL statements.

-- DANGER: The result may unexpectedly mix records from multiple indexes. Please check carefully before use. // [!code error]
-- SELECT COUNT(DISTINCT index_name) FROM vchordrq_sampled_queries WHERE ... // [!code error]

> SELECT * from vchordrq_sampled_queries;

Reviewer: 1. If it's danger, why make the color of it grey?
2. This description is confusing. Please clarify that this is only intended for debugging. Users will not understand why you are teaching them an incorrect method and even encouraging its use.

```

> To measure the recall of the recorded queries, you can reconstruct the target query statement. The `vchordrq_sampled_queries` view is **not** recommended here because the evaluated recall will be incorrect if the records come from more than one index.

Reviewer: 1. It's not `reconstruct`.
2. The text above doesn't even mention that `vchordrq_sampled_queries` is a view.
3. It's not recommended, it's strictly forbidden.

```sql
SELECT AVG(recall_value) FROM (
    SELECT vchordrq_evaluate_query_recall(

>             format(
>                 'SELECT ctid FROM %I.%I ORDER BY %I %s %s LIMIT 10',
>                 quote_ident(lq.schema_name),
>                 quote_ident(lq.table_name),
>                 quote_ident(lq.column_name),
>                 lq.operator,
>                 quote_literal(lq.value)
>             )

Reviewer: ```suggestion
            format(
                'SELECT ctid FROM %I.%I ORDER BY %I %s %L LIMIT 10',
                lq.schema_name,
                lq.table_name,
                lq.column_name,
                lq.operator,
                lq.value
            )
```

https://www.postgresql.org/docs/current/functions-string.html

Reviewer: ```sql
                CASE
                    WHEN OP.opcname LIKE '%%l2%%' THEN '<->'
                    WHEN OP.opcname LIKE '%%ip%%' THEN '<#>'
                    WHEN OP.opcname LIKE '%%cosine%%' THEN '<=>'
                    ELSE ''
                END AS operator
```

I reviewed the code again for it. Please modify the code to avoid cheap judgments.

    ) AS recall_value
    FROM vchordrq_sampled_queries('items_embedding_idx') AS lq
) AS eval_results;
```

> ::: tip IMPORTANT

Reviewer: Please don't set strange TIPs. I don't understand the trade-offs here.

> - This feature adds an extra 0.3 to 1 millisecond of latency to SELECT statements, which can be reduced by lowering the value of `vchordrq.query_sampling_max_records`.

Reviewer: When users see the documentation written this way, they immediately realize that VectorChord performed a full table scan. I am increasingly feeling that this feature is a failure. If I were a user, I would never use it.

Author: I researched similar features in the documentation to see how they describe additional latency when logging.

## log_statement
No such description.

## pgaudit
The description is very vague and requires users to take responsibility.

> Depending on settings, it is possible for pgAudit to generate an enormous volume of logging. Be careful to determine exactly what needs to be audit logged in your environment to avoid logging too much. ... Even if the insert volume is not as high, the performance impact of audit logging may still noticeably affect latency.

## pg_wait_sampling
No such description.

## auto_explain
The description is very vague and requires users to take responsibility.

> When this parameter is on, per-plan-node timing occurs for all statements executed, whether or not they run long enough to actually get logged. This can have an extremely negative impact on performance. Turning off auto_explain.log_timing ameliorates the performance cost, at the price of obtaining less information.

## Conclusion
Perhaps we should not discuss it at all, or simply say, "This can negatively impact query performance."

- Query records are not synchronized by replication or backup. Each Postgres replica maintains its own set of records based on the queries it processes.
:::

## Reference

This feature is not supported by `vchordg` and Multi-Vector Retrieval.

### Search Parameters <badge type="info" text="vchordrq" /> {#search-parameters}

#### `vchordrq.query_sampling_enable` <badge type="tip" text="since v0.5.2" />

- Description: Enables or disables the query sampling of `vchordrq` index scan.
- Type: boolean
- Default: `off`

#### `vchordrq.query_sampling_max_records` <badge type="tip" text="since v0.5.2" />

- Description: Controls the maximum number of queries recorded per index. This setting is only active when `vchordrq.query_sampling_enable` is on.
- Type: integer
- Default: `0`
- Domain: `[0, 10000]`
- Example:
    - `vchordrq.query_sampling_max_records = 0` will maintain 0 queries for each index.
    - `vchordrq.query_sampling_max_records = 1000` will maintain up to the 1000 most recent queries for each index.

#### `vchordrq.query_sampling_rate` <badge type="tip" text="since v0.5.2" />

> - Description: Controls the fraction of vchordrq index scan queries that will be recorded like [log_statement_sample_rate](https://www.postgresql.org/docs/current/runtime-config-logging.html#GUC-LOG-STATEMENT-SAMPLE-RATE). This setting is only active when `vchordrq.query_sampling_enable` is on.

Reviewer: 1. Two GUCs have nothing in common, except their names. That GUC disables the sampling while this doesn't.
2. Only superusers and users with the appropriate SET privilege can change that GUC. But it's a good idea, we should do the same thing in VectorChord.
3. We'd better clarify that this is random, rather than referring to external documents and let users guess what we want them to look at.

- Type: real
- Default: `0`
- Domain: `[0, 1]`
- Example:
    - `vchordrq.query_sampling_rate = 0` records no query.
    - `vchordrq.query_sampling_rate = 0.01` records approximately 1% of queries.
    - `vchordrq.query_sampling_rate = 1` records every query.

### Functions <badge type="info" text="vchordrq" />

#### `vchordrq_sampled_queries` <badge type="tip" text="since v0.5.2" /> {#vchordrq-sampled-queries-func}

> - Description: Retrieves the set of recorded queries for a **single**, **specified** vchordrq index.

Reviewer: ```suggestion
- Description: Retrieves the set of recorded queries for a `vchordrq` index.
```

- Result: `table`
    - `schema_name: text`, the schema of the table in the query
    - `index_name: text`, the name of the index used in the query
    - `table_name: text`, the name of the table in the query
    - `column_name: text`, the column involved in the index scan
    - `operator: text`, the distance [operator](/vectorchord/usage/indexing#operator-classes) used in the query.
    - `value: text`, the text representation of the target vector from the query
- Arguments:
    - `regclass`, the object identifier of a `vchordrq` index
- Example:

>     - `SELECT vchordrq_sampled_queries('idx');`

Reviewer: ```suggestion
    - `SELECT vchordrq_sampled_queries('items_embedding_idx');`
```

    - `SELECT vchordrq_evaluate_query_recall(query => $$ SELECT ctid FROM t ORDER BY val <-> '[0.5, 0.25, 1.0]' LIMIT 10 $$);`
    - `SELECT vchordrq_evaluate_query_recall(query => $$ SELECT ctid FROM t ORDER BY val <-> '[0.5, 0.25, 1.0]' LIMIT 10 $$, exact_search => true);`
    - `SELECT vchordrq_evaluate_query_recall(query => $$ SELECT ctid FROM t ORDER BY val <-> '[0.5, 0.25, 1.0]' LIMIT 10 $$, accu_probes => '9999', accu_epsilon => 4.0);`

### Views <badge type="info" text="vchordrq" />

#### `vchordrq_sampled_queries` <badge type="tip" text="since v0.5.2" />

> - Description: Retrieves the set of recorded queries for **all** vchordrq indexes.

Reviewer: Please clarify that this is only intended for debugging.
